### PR TITLE
[ACCESSIBILITY] Global changes to pass AXE/Lighthouse Audits

### PR DESCRIPTION
**Pull request by austin94**
**TL;DR**: We have improved from an aggregate lighthouse accessibility score of 45% to 98.5% across **ALL** packages (with the remaining failures being lighthouse miss-fires) 🎉 

* [X] **BREAKING CHANGES?**

⚠️ There are several breaking changes in this PR. This is just a collection of all changes that I completed. This should (and will) be broken up into several smaller PRs that can address the changes in more depth. ⚠️ 

## Tooling Used
- [Google Lighthouse Audits](https://developers.google.com/web/tools/lighthouse/)
  - The google Lighthouse accessibility audit is a good first-pass for accessibility health
- [AXE Auditing Tool](https://www.deque.com/axe/)
  - Chrome extension for more involved accessibility testing
- [Mac Voiceover](https://www.apple.com/accessibility/mac/vision/)
  - Nothing beats actually testing with a screen reader

All three tools were used for the tests/assumptions performed during this process.

Here is a link to the lighthouse exports after the changes shown in this PR: https://drive.google.com/open?id=11bOPsxBof8gMa8Md0ZXhKzvMg0PHyVYq

These files can be viewed with https://googlechrome.github.io/lighthouse/viewer2x/

## Description

Accessibility is important.  All components within this repository are built to the [W3C Design Patterns](https://www.w3.org/TR/wai-aria-practices/#aria_ex) spec; so this means that everything is 100% accessible, right?

**WRONG** 😢 (this is [react-buttons](https://garden.zendesk.com/react-components/buttons/)) 

old-lighthouse-audit

The majority of component implementations are accessible, but our examples are currently lacking some good best practices. Some of this is a combination of color-contrast issues and customizations on top of react-styleguidist, but the worst offenders currently are the handwritten examples and some invalid assumptions about menu accessibility.

Currently all examples are hovering around the 40-60% lighthouse score.  This PR is a collection of several breaking API changes, styleguidist improvements, and example updates. These will be split into several individual PRs, but I felt a background PR would help.

## Detail

I've split this PR into 3 commits:

* Breaking (and non-breaking) API Changes
  * Turns out sticking to the W3C spec doesn't provide _everything_. I'll go into depth in the individual PRs
  * Examples:
    * Can't apply two IDs to `aria-describedby`
    * `aria-controls` and `aria-owns` can only to reference elements that are rendered already (duh)
    * .etc

* Example updates
  * Assorted `aria-label` and documentation changes

* Styleguidist upgrades
  * Tried to use Garden colors throughout
  * Migrated propTypes table to use the Garden table component

Before:
old

After:
new

## Learnings

Even thought we have static accessibility linting with`jsx-a11y` a large chunk of our codebase lives within the Markdown examples and isn't reachable with this tool.  We should perform more frequent accessibility audits or find some additional ways of auditing this in an automated fashion.

## Checklist

* [x] :ok_hand: design updates are Garden Designer approved (add the
      designer as a reviewer)
* [ ] :nail_care: view component styling is based on a Garden CSS
  component
* [ ] :globe_with_meridians: Styleguidist demo is up-to-date (`yarn start`)
* [ ] :arrow_left: renders as expected with reversed (RTL) direction
* [ ] :guardsman: includes new unit and snapshot tests
* [ ] :ledger: any new files are included in the packages `src/index.js` export
* [ ] :memo: tested in Chrome, Firefox, Safari, Edge, and IE11

### CODE REVIEW

**Comment by allisonacs** [comment]
Happy to see this happen @austin94! Less "do as I say," more "do as I do."

**Comment by austin94** [comment]
> Nice, so these all came out of running the styleguide through the chrome dev tools audit panel

Unfortunately no. The overall scores did, but the majority of corrections were guided by the AXE plugin. It seems to be a lot more picky about things than the lighthouse audit and helped find some of the larger holes.

File: packages/autocomplete/src/containers/AutocompleteContainer.js [hunk 1]
```diff
@@ -340,7 +338,6 @@ class AutocompleteContainer extends ControlledComponent {
       role,
       'aria-autocomplete': 'list',
       'aria-haspopup': 'true',
-      'aria-owns': this.getMenuId(),
```
[the remark this answers is not in the document]
Comment (austin94) [reply packages/autocomplete/src/containers/AutocompleteContainer.js]: No it does not. Since we are moving focus to the menu it seems to not be required? I ran AXE across all of the states in the menu and the only thing it picked up was us appending it to the body in a "non-landmark region" which would be something a consumer would customize.

File: packages/selection/src/containers/FieldContainer.spec.js [hunk 2]
```diff
@@ -48,15 +48,26 @@ describe('FieldContainer', () => {
...
+        <FieldContainer id={CONTAINER_ID}>
+          {({ getInputProps }) => (
+            <div>
+              <input {...getInputProps({ 'data-test-id': 'input' }, { isDescribed: false })} />
+            </div>
+          )}
+        </FieldContainer>
       );
+
+      expect(findInput(wrapper)).toHaveProp('aria-describedby', null);
```
[the remark this answers is not in the document]
Comment (austin94) [reply packages/selection/src/containers/FieldContainer.spec.js]: Unfortunately no. The syntax above would expect the prop to have `undefined` as a value, but is null as provided.

If we were to provide `undefined` through the render-prop it would actually show as `undefined` in the DOM. With `null` React is removing it from the attribute list.

File: packages/selection/src/containers/FieldContainer.spec.js [hunk 3]
```diff
@@ -48,15 +48,26 @@ describe('FieldContainer', () => {
...
 
       expect(input).toHaveProp('id', `${CONTAINER_ID}--input`);
       expect(input).toHaveProp('aria-labelledby', `${CONTAINER_ID}--label`);
-      expect(input).toHaveProp(
-        'aria-describedby',
-        `${CONTAINER_ID}--hint ${CONTAINER_ID}--message`
+      expect(input).toHaveProp('aria-describedby', `${CONTAINER_ID}--hint`);
+    });
+
+    it('excludes aria-describedby if option is provided', () => {
```
[the remark this answers is not in the document]
Comment (austin94) [reply packages/selection/src/containers/FieldContainer.spec.js]: Fixed. Was going to save this for the separate PR, but depending on what we view as the most common scenario, we might have this default to false.

Updated for now.

File: packages/radios/src/views/Label.example.md [hunk 4]
```diff
@@ -1,38 +1,40 @@
...
-      </Radio>
-    </Col>
-    <Col md={4}>
-      <Radio disabled>
-        <Label>Disabled Label</Label>
-      </Radio>
-    </Col>
-  </Row>
-</Grid>
+<div role="group" aria-label="State Usage Example">
```
Comment (jzempel) [inline packages/radios/src/views/Label.example.md]: Can these attributes apply to `<Grid>`? In other words, make `role="group"` automatic with the component and apply `aria-label` manually.

File: packages/tables/src/examples/overflow-menu.md [hunk 5]
```diff
@@ -31,7 +34,7 @@ const OverflowMenu = ({ isHeader = false }) => (
         enabled: false
       },
       offset: {
-        fn: (data) => {
+        fn: data => {
```
Comment (jzempel) [inline packages/tables/src/examples/overflow-menu.md]: hopefully grabbed by rebase on #65

File: packages/textfields/src/views/__snapshots__/Input.spec.js.snap [hunk 6]
```diff
@@ -2,6 +2,7 @@
 
 exports[`Input renders RTL styling 1`] = `
 <input
+  aria-invalid={false}
```
Comment (jzempel) [inline packages/textfields/src/views/__snapshots__/Input.spec.js.snap]: Isn't `{false}` in these (and following) snaps unexpected?

File: utils/styleguide/ArgumentsRenderer/index.js [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+ */
+
+import React from 'react';
+import PropTypes from 'prop-types';
+import styled from 'styled-components';
+import Argument from 'react-styleguidist/lib/rsg-components/Argument/ArgumentRenderer';
+import { zdFontSizeEpsilon, zdFontWeightSemibold } from '@zendeskgarden/css-variables';
+
+const ArgumentsContainer = styled.div`
+  margin-bottom: 8px;
```
Comment (jzempel) [inline utils/styleguide/ArgumentsRenderer/index.js]: Use `xs` spacing from css-variables.

File: utils/styleguide/ArgumentsRenderer/index.js [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+  font-size: inherit;
+`;
+
+const Heading = styled.div`
+  margin-bottom: 4px;
```
Comment (jzempel) [inline utils/styleguide/ArgumentsRenderer/index.js]: Use `xxs` spacing from css-variables.

File: utils/styleguide/github-hljs-theme.css [hunk 9]
```diff
@@ -0,0 +1,90 @@
+.hljs {
```
Comment (jzempel) [inline utils/styleguide/github-hljs-theme.css]: If you can't access variables, it would be good to comment this file with selected color names from https://garden.zendesk.com/css-components/utilities/color/.

File: packages/radios/src/views/Label.example.md (hunk 4 above)
Comment (austin94) [reply packages/radios/src/views/Label.example.md]: These are no added to the `<Grid>`

File: packages/tables/src/examples/overflow-menu.md (hunk 5 above)
Comment (austin94) [reply packages/tables/src/examples/overflow-menu.md]: Fixed with a rebase

File: packages/textfields/src/views/__snapshots__/Input.spec.js.snap (hunk 6 above)
Comment (austin94) [reply packages/textfields/src/views/__snapshots__/Input.spec.js.snap]: We are defaulting this to `false` to match the "default" state (https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques/Using_the_aria-invalid_attribute).

Without any validation this would still be false.  AXE didn't seem to find any issues with this.

File: utils/styleguide/ArgumentsRenderer/index.js (hunk 7 above)
Comment (austin94) [reply utils/styleguide/ArgumentsRenderer/index.js]: Fixed

File: utils/styleguide/ArgumentsRenderer/index.js (hunk 8 above)
Comment (austin94) [reply utils/styleguide/ArgumentsRenderer/index.js]: Fixed

File: utils/styleguide/github-hljs-theme.css (hunk 9 above)
Comment (austin94) [reply utils/styleguide/github-hljs-theme.css]: I've added the equivalent color values as comments. I also updated all of the colors to match an equivalent in the palette so it should be a little clearer now.

File: packages/textfields/src/views/__snapshots__/Input.spec.js.snap (hunk 6 above)
Comment (jzempel) [reply packages/textfields/src/views/__snapshots__/Input.spec.js.snap]: But shouldn't it be `aria-invalid="false"` rather than `{false}`? I haven't seen JSX syntax in snapshots before.

Comment (jzempel) [reply packages/textfields/src/views/__snapshots__/Input.spec.js.snap]: nvmd. Bogus expectations on my part here. 😳

**Comment by ryanseddon** [comment]
> AXE plugin

Oh weird I thought from [this post](https://www.deque.com/blog/google-selects-deques-axe-chrome-devtools/) that lighthouse actually internally uses AXE in the audit?

**Comment by austin94** [comment]
> Oh weird I thought from this post that lighthouse actually internally uses AXE in the audit?

I think it does, it at least linked to the dequeue links that AXE provides, but it seems to be a sub-set of the audits that the standalone plugin uses.

> So next steps is to introduce smaller PRs for these changes?

Yes.  Probably 3-4

**Comment by austin94** [comment]
Closing this now that all areas have been merged separately.
